File: projects/tax-id/src/lib/countries/non-european-vat.ts
import { TaxIdValidationResult } from '../models';

// UAE Tax Registration Number: the Federal Tax Authority verifier requires a
// 15-character TRN. VAT TRNs are numeric; issuance/activity still requires
// the authoritative online verifier, so this rule is format-only.
export function validateEmiratiVat(value: unknown): TaxIdValidationResult {
  const normalizedValue =
    typeof value === 'string' || typeof value === 'number'
      ? String(value).trim().replace(/[\s-]+/g, '')
      : '';
  const base = { country: 'AE', normalizedValue } as const;

  if (!normalizedValue) return { ...base, valid: false, error: 'empty' };
  if (normalizedValue.length !== 15) {
    return { ...base, valid: false, error: 'invalid_length' };
  }
  if (!/^\d{15}$/.test(normalizedValue) || normalizedValue === '000000000000000') {
    return { ...base, valid: false, error: 'invalid_format' };
  }
  return { ...base, valid: true, validationLevel: 'format' };
}

File: packages/dotnet/NationalIdentifiers.Core/Countries/NonEuropeanVat.cs
using System.Text.RegularExpressions;

namespace NationalIdentifiers.Core.Countries;

internal static class NonEuropeanVat
{
    // UAE Tax Registration Number: the Federal Tax Authority verifier requires
    // 15 characters. VAT TRNs are numeric; issuance/activity requires the
    // authoritative online verifier, so this rule is format-only.
    internal static ValidationResult UnitedArabEmirates(object? value)
    {
        var normalized = value switch
        {
            string s => Regex.Replace(s.Trim(), @"[\s-]+", string.Empty),
            int or long or double or float or decimal => value.ToString() ?? string.Empty,
            _ => string.Empty,
        };

        if (string.IsNullOrEmpty(normalized))
            return ValidationResult.Fail("AE", normalized, ValidationErrorCode.Empty);
        if (normalized.Length != 15)
            return ValidationResult.Fail("AE", normalized, ValidationErrorCode.InvalidLength);
        if (!Regex.IsMatch(normalized, @"^\d{15}$") || normalized == "000000000000000")
            return ValidationResult.Fail("AE", normalized, ValidationErrorCode.InvalidFormat);
        return ValidationResult.Ok("AE", normalized, ValidationLevel.Format);
    }
}
